[SirMentos] iP

SSirMentos opened the pull request:
# MentosBot - Task Management System

_"Stay fresh, stay organized."_ – MentosBot 🧠

MentosBot is designed to help you manage your tasks effortlessly. It’s:
* **text-based**
* **easy to learn**
* **FAST** – **SUPER FAST** to use

## Getting Started
Follow these simple steps to get started:

1. [Download MentosBot](https://github.com/SSirMentos/ip) from the repository.
2. Double-click the downloaded file.
3. Add your tasks using simple commands.
4. Let MentosBot handle the rest for you! 😉

## Features
* **Task management**: Create, update, delete tasks.
* **Event and deadline tracking**: Keep track of important dates and events.
* **Search functionality**: Quickly find tasks by keywords.
* **Reminders**: *(coming soon)*

## Usage Example

Here's how you can start the application:

```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SSirMentos committed:
Level 0. Rename, Greet, Exit to Mentos

SSirMentos committed:
Added echoBack function so that the bot will echo whatever things user input except bye

SSirMentos committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

SSirMentos committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

SSirMentos committed:
Merge branch 'master' of https://github.com/SSirMentos/ip

* 'master' of https://github.com/SSirMentos/ip:
  Add the ability to store whatever text entered by the user and display them back to the user when requested

# Conflicts:
#	src/main/java/Mentos.java

SSirMentos committed:
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

SSirMentos committed:
test

SSirMentos committed:
Add support for tracking three types of tasks:

ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

SSirMentos committed:
Use the input/output redirection technique to semi-automate the testing of Mentos.

SSirMentos committed:
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

SSirMentos committed:
Add support for deleting tasks from the list.

SSirMentos committed:
Added Javadocs

SSirMentos committed:
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up.

SSirMentos committed:
Teach the chatbot how to understand dates and times. For example, if the command is deadline return book /by 2/12/2019 1800, the chatbot should understand 2/12/2019 1800 as 2nd of December 2019, 6pm, instead of treating it as just a String.

SSirMentos committed:
Refactor the code to extract out closely related code as classes.

Minimal: Extract the following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list

SSirMentos committed:
remove redundant lines

SSirMentos committed:
remove additional redundant lines

SSirMentos committed:
Organize the classes into suitable java packages.

SSirMentos committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

SSirMentos committed:
configured build gradle

SSirMentos committed:
Add JUnit tests to test the behavior of the code.

SSirMentos committed:
Package the app as an executable JAR file so that it can be distributed easily.

SSirMentos committed:
Fix Exception String

SSirMentos committed:
Added javadocs

SSirMentos committed:
enforce coding standard

SSirMentos committed:
Merge tag 'A-CodingStandard'

* tag 'A-CodingStandard':
  enforce coding standard

SSirMentos committed:
Give users a way to find a task by searching for a keyword in the task description.

LiuYJ2002 commented on src/main/java/Mentos/Mentos.java:
```diff
@@ -0,0 +1,180 @@
+    private final Storage storage;
+    private final Ui ui;
+
+    public Mentos(String FILE_PATH){
```
Should there be a space before the brackets for the constructor?

LiuYJ2002 commented on src/main/java/Mentos/Mentos.java:
```diff
@@ -0,0 +1,180 @@
+    public Mentos(String FILE_PATH){
+        this.storage = new Storage(FILE_PATH);
+        this.ui = new Ui();
+        try{
```
Should there be a space before the bracket for the try-catch?

LiuYJ2002 commented on src/main/java/Mentos/Mentos.java:
```diff
@@ -0,0 +1,180 @@
+     * @throws MentosException if the index is out of bounds or zero.
+     */
+    public boolean checkIndex(int index) throws MentosException {
+        if (index > tasks.size() || index == 0){
```
Should there be a space before the bracket for the if block

LiuYJ2002 commented on src/main/java/Mentos/components/ActionTaskIndexTuple.java:
```diff
@@ -0,0 +1,28 @@
+package Mentos.components;
```
Should the package naming be in all lowercase?

LiuYJ2002 commented on src/main/java/Mentos/task/Event.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public String toString(){
+        return String.format("[E] %s (from: %s to: %s)",super.toString(),this.from.format(DateTimeFormatter.ofPattern("MMM d yyyy, HH:mm")),this.to.format(DateTimeFormatter.ofPattern("MMM d yyyy, HH:mm")));
```
Should the line length be shorter with better indentation for next line?

LiuYJ2002 reviewed (commented):
LGTM. Just some nits like spacing before brackets are quite common.

ywllowsensor commented on src/main/java/Mentos/components/ActionTaskIndexTuple.java:
```diff
@@ -0,0 +1,28 @@
+
+import Mentos.task.Task;
+
+public class ActionTaskIndexTuple {
```
At first glance, the class name here seems slightly complicated to me! Perhaps a header comment could be included here to explain what the class does? Alternatively, you can also consider using a shorter and simpler name!

ywllowsensor commented on src/main/java/Mentos/task/Event.java:
```diff
@@ -0,0 +1,21 @@
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task{
+    protected LocalDateTime from;
+    protected LocalDateTime to;
```
Since LocalDateTime is a noun, perhaps you can consider renaming these fields such that they sound more like nouns! (eg. fromDate or toDate)

ywllowsensor commented on src/main/java/Mentos/components/Ui.java:
```diff
@@ -0,0 +1,155 @@
+     * @param task The task that has been unmarked as not done.
+     */
+    public void unmarkEvent(Task task){
+        System.out.println("Holdup this Mentos.task is not done!");
```
I like the little bits of personalisation that you have added to your ChatBot. Great work!

ywllowsensor commented on src/main/java/Mentos/components/Ui.java:
```diff
@@ -0,0 +1,155 @@
+     * @param tasksSize The number of tasks remaining after the deletion.
+     */
+
+    public void deleteEvent(Task task, int tasksSize){
```
The name of this method could be slightly misleading as the Javadoc comment mentions that it is deleting a task. But the method is named "deleteEvent". You can consider changing its name to something like "deleteTask"!

ywllowsensor reviewed (commented):
Overall, the code looks good to me! I can see that you are still in the midst of implementing improvements such as inclusion of the Javadoc comments. Keep up the work!

SSirMentos committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

SSirMentos committed:
Add assertions to document key assumptions in Mentos class

- Added `assert` statements to ensure valid task indices during task operations (mark, unmark, delete,searchTaskGUI).
- Validated that tasks are loaded correctly from storage with assertions.
- Ensured non-null and valid task objects before adding, marking, or unmarking tasks.

SSirMentos committed:
Refactor command handling to an OOP structure

- Converted individual commands (List, Mark, Unmark, Delete, ToDo, Deadline, Event, Find) into their own classes.
- Improved task handling logic by leveraging polymorphism and encapsulation for better maintainability.
- Replaced the if-else structure with an object-oriented design, promoting cleaner code and separation of concerns.
- Simplified the `taskHandler` method to dynamically handle commands through respective command objects.

SSirMentos committed:
Update test cases for assertions

SSirMentos committed:
Add support for editing details of existing items

- Implemented functionality to allow users to update specific details of items (e.g., changing the end time of an event) without deleting the item.
- Enabled editing of task details (e.g., description, date) while preserving other information.

SSirMentos committed:
Deleted the ActionTaskTuple class as it was no longer being used in the codebase.

SSirMentos committed:
Merge pull request #1 from SSirMentos/BCD-Extension-Update

Deleted the ActionTaskTuple class as it was no longer being used in t…

SSirMentos committed:
Improve Mentos GUI

SSirMentos committed:
Add date validation for Event Class
Add a more robust regex for ToDo

SSirMentos committed:
Update Readme.md and MentosBot avatar

SSirMentos committed:
Fixed Update bugs and Updated Readme

SSirMentos committed:
Update Readme

SSirMentos committed:
Clean Storage code

SSirMentos committed:
Handle empty input

SSirMentos committed:
Update README.md

SSirMentos committed:
Add Javadocs to Event.java

SSirMentos committed:
Add Javadocs to Command.java

SSirMentos committed:
Add Javadocs to Task.java

SSirMentos committed:
Add Javadocs to Deadline.java
